clang/lib/Sema/SemaExpr.cpp

```diff
@@ -11444,9 +11444,12 @@ static void DiagnoseBadShiftValues(Sema& S, ExprResult &LHS, ExprResult &RHS,
   llvm::APSInt Right = RHSResult.Val.getInt();
 
   if (Right.isNegative()) {
-    S.DiagRuntimeBehavior(Loc, RHS.get(),
-                          S.PDiag(diag::warn_shift_negative)
-                            << RHS.get()->getSourceRange());
+    if (S.ExprEvalContexts.back().isConstantEvaluated())
+      S.Diag(Loc, diag::warn_shift_negative) << RHS.get()->getSourceRange();
+    else
+      S.DiagRuntimeBehavior(Loc, RHS.get(),
+                            S.PDiag(diag::warn_shift_negative)
+                              << RHS.get()->getSourceRange());
```

Comment:
I still don't think this is the correct fix -- I think the issue is that ExprConstant.cpp needs to be taught about these as undefined behaviors. Notice how GCC correctly emits the pedantic diagnostic but Clang does not: https://godbolt.org/z/arb8Y9G9q

When I step through in the debugger, it seems that we do catch the issue in `handleIntIntBinOp()` but we're suppressing diagnostics for it in C more broadly. That seems more likely to be the root cause of the issue. (Basically, we want the diagnostics to be split between SemaExpr.cpp for runtime reachable diagnostics and ExprConstant.cpp for constant expressions.)

CC @tbaederr for opinions on constant expression behavior here.

Reply:
The constant interpreter already diagnoses this:
https://github.com/llvm/llvm-project/blob/51089e360e37962c7841fe0a494ba9fb5368bab2/clang/lib/AST/ExprConstant.cpp#L2850-L2851

and for the test case with the `(-1 << 29)` enum value, this is evaluated in `Sema::VerifyIntegerConstantExpresssion()` and it's also diagnosed here:
https://github.com/llvm/llvm-project/blob/51089e360e37962c7841fe0a494ba9fb5368bab2/clang/lib/Sema/SemaExpr.cpp#L18158-L18159

But the diagnostic doesn't show up. Not sure why. If I modify the constant interpreter to `return false` in the negative LHS case, I get the expected diagnostic:
```console
/home/tbaeder/test.cpp:2:9: error: expression is not an integral constant expression
    2 |     X = (-1<<29) // expected-warning {{shifting a negative signed value is undefined}}
      |         ^~~~~~~~
/home/tbaeder/test.cpp:2:12: note: left shift of negative value -1
    2 |     X = (-1<<29) // expected-warning {{shifting a negative signed value is undefined}}
      |            ^
```

Reply:
Diagnose is caught in handleIntIntBinOp() and is emitted for C++ in VerifyIntegerConstantExpression(). There are two reasons why diagnostic could be emitted in case of C++:
	1) evaluating expression as rvalue fails or evaluation result has side effects or it is not an integer; evaluating expression as rvalue fails if function handleIntIntBinOp() returns false
	2) folding is allowed (CanFold == AllowFoldKind::AllowFold); this value depends on place where constant expression is defined; eg. in case on enum folding is allowed and in case of static assertion it is not allowed
Also, diagnostic would not be emitted if Diagnoser is suppresed.

However, in C case there are no diagnostic for integer constant expression.

So I enabled diagnosing in C case (if Diagnoser is not suppressed). In function EvaluateKnownConstIntCheckOverflow() it is expected that evaluating rvalue is successful, so I keep handleIntIntBinOp() returning true.
For C++ diagnostic, I modified handleIntIntBinOp() to return false.

As folding is allowed for enum and not for static assertion, I added in tests to check both enum and static assertion diagnosing.